### PR TITLE
[WIP] Fix test fail without mps

### PR DESCRIPTION
**Pull request by yuhaoth**
This is try to fix test fail without mps . That is mentioned in #238 . 

I am not sure if it is useful for #238

### CODE REVIEW

File: include/mbedtls/ssl.h [hunk 1]
```diff
@@ -670,9 +635,27 @@ typedef enum
...
+    MBEDTLS_SSL_CLIENT_NEW_SESSION_TICKET,
+#if defined(MBEDTLS_SSL_TLS13_COMPATIBILITY_MODE)
+    MBEDTLS_SSL_CLIENT_CCS_BEFORE_2ND_CLIENT_HELLO,
+    MBEDTLS_SSL_CLIENT_CCS_AFTER_SERVER_FINISHED,
+    MBEDTLS_SSL_CLIENT_CCS_AFTER_CLIENT_HELLO,
+    MBEDTLS_SSL_SERVER_CCS_AFTER_SERVER_HELLO,
+    MBEDTLS_SSL_SERVER_CCS_AFTER_HRR,
+#endif /* MBEDTLS_SSL_TLS13_COMPATIBILITY_MODE */
+    MBEDTLS_SSL_EARLY_APP_DATA
+#endif
```
Comment (zhihan) [inline include/mbedtls/ssl.h]: / * MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */

File: library/ssl_msg.c [hunk 2]
```diff
@@ -2984,7 +2984,8 @@ int mbedtls_ssl_write_handshake_msg_ext( mbedtls_ssl_context *ssl,
         if( ( ret = mbedtls_ssl_write_record( ssl, SSL_FORCE_FLUSH ) ) != 0 )
         {
             MBEDTLS_SSL_DEBUG_RET( 1, "ssl_write_record", ret );
-            return( ret );
+            if(ret != MBEDTLS_ERR_SSL_WANT_WRITE)
+                return( ret );
```
Comment (zhihan) [inline library/ssl_msg.c:2988]: Trying to understand, if we do not return here, we would return on line 2994 with 0, right? Would that confuse the caller of this function to wrongfully assume the write has been successful?

**Comment by lhuang04** [comment]
@yuhaoth 
Thanks for fixing. The fix around `mbedtls_ssl_fetch_input` is definitely helpful for #238. However other issues in #238 may need fix too:
* We always [send alert](https://github.com/hannestschofenig/mbedtls/blob/tls13-prototype/library/ssl_tls.c#L6863) when the ret from `mbedtls_ssl_handshake_step` is non zero.
* When `MBEDTLS_ERR_SSL_WANT_WRITE` returned from [mbedtls_ssl_write_handshake_msg](https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/xplat/mobilenetwork/third-party/mbedtls/library/ssl_tls13_server.c?commit=7f447c68e2784949de947ef749e6dac59bf2b672&lines=3842) or [mbedtls_ssl_flush_output](https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/xplat/mobilenetwork/third-party/mbedtls/library/ssl_tls13_server.c?commit=2165162655f56181af41b19ed91f34c2e2a3d8e9&lines=4284), there is no guarantee that we could re-enter the state machine seamless.  We often stuck in the same state or repeat the calculation.

File: library/ssl_msg.c (hunk 2 above)
Comment (lhuang04) [reply library/ssl_msg.c:2988]: Yeah, we may need propagate `MBEDTLS_ERR_SSL_WANT_WRITE` to upper level.

File: library/ssl_msg.c [hunk 3]
```diff
@@ -3179,7 +3180,8 @@ int mbedtls_ssl_write_record( mbedtls_ssl_context *ssl, uint8_t force_flush )
         ( ret = mbedtls_ssl_flush_output( ssl ) ) != 0 )
     {
         MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_flush_output", ret );
-        return( ret );
+        if(ret != MBEDTLS_ERR_SSL_WANT_WRITE)
```
Comment (lhuang04) [inline library/ssl_msg.c:3183]: same here, we may need propagate MBEDTLS_ERR_SSL_WANT_WRITE to upper level

File: library/ssl_msg.c (hunk 2 above)
Comment (hanno-becker) [reply library/ssl_msg.c:2988]: I agree, we can't ignore the error. @yuhaoth What was the motivation behind this change?

File: library/ssl_msg.c (hunk 3 above)
Comment (hanno-becker) [reply library/ssl_msg.c:3183]: Agree

**Comment by hanno-becker** [comment]
Changed target branch back and forth to make github realize that some base PRs have been merged by now.

**Review by hanno-becker — CHANGES_REQUESTED** [review]
Thank you for looking into this, @yuhaoth.

As @lhuang04 and @zhihan mentioned, we can not ignore `WANT_READ/WRITE` error codes. What needs to understood to fix #291 is where, in the non-MPS code, we're running into those error codes and return in a non-reentrant way. This is what needs to be fixed.

File: library/ssl_msg.c (hunk 2 above)
Comment (yuhaoth) [reply library/ssl_msg.c:2988]: As @lhuang04  said , propagate it  to upper level .

**Comment by hanno-becker** [comment]
This has been fixed in #346

**Pull request closed without merging**